### ISSUE TITLE
[security] .envrc file permissions: Is it OK to allow .envrc to be writeable by other users?

### ISSUE THREAD

**Issue by pmav99** [direnv/direnv#445]
## Issue description

As @asottile shows [here](https://github.com/Yelp/aactivator/issues/29) direnv is happy to allow `.envrc` to be writeable by other users. This permits a malicious user to e.g. change the contents of `.envrc` right after you inspected the contents of `.envrc` and just before you run `direnv allow`.

I can understand that in some cases you might find it convenient to e.g. have an `.envrc` that is writable by group members. But if that is a desired feature, as far as security goes, wouldn't it be better to only permit user-only writeable `.envrcs` and only allow more liberal permissions if the user explicitly allows them?

For convenience, I am copy pasting the relevant parts of @asottile 's comments

Initial setup:
``` bash
$ eval "$(~/go/bin/direnv hook bash)"
$ type cd
cd is a shell builtin
$ echo "$PROMPT_COMMAND"
_direnv_hook;eval "`/home/asottile/opt/venv/bin/aactivator`"; if [ -d .git -a ! -x .git/hooks/pre-commit -a -e .pre-commit-config.yaml ] && which pre-commit >& /dev/null; then pre-commit install; fi; 
$ mkdir /tmp/foo
$ chmod 777 /tmp/foo
$ echo 'export FOO=1' > /tmp/foo/.envrc
$ ~/go/bin/direnv allow /tmp/foo/
$ cd /tmp/foo/
direnv: loading .envrc
direnv: export +FOO
```

malicious user:
``` bash
$ cat /tmp/foo/.envrc 
FOO=1
$ echo looks good to me'!'
looks good to me!
$ # simulate a malicious actor
$ sudo -u nobody bash -c 'echo "export FOO=haxxed" > /tmp/.envrc && mv -f /tmp/.envrc /tmp/foo'
$ ~/go/bin/direnv allow /tmp/foo/
$ cd /tmp/foo/
direnv: loading .envrc
direnv: export +FOO
$ echo $FOO
haxxed
```

**Comment by daniels** [comment]
Isn't this risk the same with any script you download, inspect and run?

**Comment by asottile** [comment]
@daniels no, this is a situation where _another user_ has write permission to _your_ files (~usually not the case -- though there are some situations where it happens)

**Comment by asottile** [comment]
The computer provides mechanisms to protect from this (the writable bit). In the demo above, `sudo` is only used to **downgrade** permissions (to `nobody` which conventionally has very little to no permissions). The security check that's being asked for here is pretty simple and unlikely to affect normal execution while patching potentially malicious users and unsafe behavior.

The minimum to implement this is to traverse up from the current directory and make sure none of the parts are global or group writable (otherwise one who is not you could modify them **without root**)

Of course all bets are off if an attacker has gained root permissions, but that's not what this issue is about at all.

**Comment by pmav99** [comment]
I am pretty sure that everybody agrees that:

- If somebody gains access to the root account, then he can do pretty much anything.
- If somebody gains access to the account using direnv, then he can do a lot of things the least of which is editing `.envrc` and issuing `direnv allow`.

But let's say that there is a system with two non-root accounts, A and B. If both A and B can edit `.envrc` then, someone gaining access to account B [can wait](https://superuser.com/questions/181517/how-to-execute-a-command-whenever-a-file-changes) until A edits `envrc` and then, when A issues `direnv allow`, account B can pretty much impersonate as account A.

I am not saying this is a huge attack vector, it is actually rather small, but it is there.

**Comment by zimbatm** [comment]
What if A and B want to collaborate on the same filesystem? I am not convinced that preventing direnv to work in that scenario would not prevent some legitimate, and already deployed, use-cases.

Similar arguments can also be applied to any program that has startup scripts in it's configuration. Or any code base's build script. Any build system config files that can be mutated can be changed to anything.

Another attack vector is to create a repository with the .envrc loading any other file. Since direnv only checks the .envrc, the user won't be asked for configurmation when that other file changes either locally or on `git pull`.

**Comment by daniels** [comment]
> @daniels no, this is a situation where _another user_ has write permission to _your_ files (~usually not the case -- though there are some situations where it happens)

Yes, but that situation is user imposed, not something direnv has setup for you. In the example above the user explicitly demands that the directory should be editable by untrusted users. Direnv allows it, but so does the OS for any executable in that directory.

You could argue that direnv should check the permissions in this specific case *just because it could*, but if so you should also require that same check from any other utility that executes commands from a different file, like `sh` or `ruby` etc.

> > Another attack vector is to create a repository with the .envrc loading any other file. Since direnv only checks the .envrc, the user won't be asked for configurmation when that other file changes either locally or on git pull.
> 
> I don't think that vector is valid, `direnv` hashes the contents of the file and asks for a new "allow" whenever they change I believe

I guess zimbatm would know if that was the case ... But it's clearly stated in the documentation that only .envrc is checked. (I think it would be great to be able to specify additional files to be checked, but I guess that would be a different issue.)

**Comment by matu3ba** [comment]
@pmav99 
You are basically asking for full user-space isolation, which is either very complex  (see firejail) or must be rigerous from the start (bubblewrap) with disallowing everything not explicitly allowed.

Note, that even though firejail tries hard, for example access control in X11 comes at a high performance cost due to the non-existing underlying security model.

Further note, that sandboxing programs intended to configure, build and run other programs in a potentially very complex desktop requires more complexity than firejail implements with more potential bugs=CVEs in the solution.

In summary, your request is unviable to implement for all use cases direnv supports, except that direnv could provide some links for users how to setup (basic) isolation themself, if direnv wants to do the maintenance.

**Comment by pmav99** [comment]
No, I didn't ask for full user isolation. This would obviously be completely out of scope for direnv. That being said, if there was a good solution for that, then a direnv-like project could consider leveraging on it.

I only asked whether direnv would be interested in requiring more restrictive file permissions. E.g. something similar to the permissions requirements of openssh (0700 for `~/.ssh/` and 0600 for the keys).

**Comment by matu3ba** [comment]
> That being said, if there was a good solution for that, then a direnv-like project could consider leveraging on it.

Unixoids have no capability based permissions and only simple option I am aware of is based on linux, meaning https://github.com/containers/bubblewrap or firejail [include ip table support] or the program permissions would need to be finegrainely provided.
Ambigiously owner filepaths may produce bad consequences, like expected stuff not running.

Afaiu, the nix sandbox is only active during builds and a simple chroot (not sure how deactivating network works).
